Composer/packages/tools/language-servers/language-generation/src/LGServer.ts

```diff
@@ -0,0 +1,355 @@
+    /**
+     * - Hi, @{name}, what's the meaning of 'state'
+     * - Hi---------, @{name}--------, what-------' ------s the meaning of "state"
+     * - <plaintext>, @{<expression>}, <plaintext><single><plaintext>------<double>
+     */
+    let i = 0;
+    while (i < lineContent.length) {
+      const char = lineContent.charAt(i);
+      if (char === `'`) {
+        if (state[state.length - 1] === 'expression' || state[state.length - 1] === 'double') {
+          state.push('single');
+        } else {
+          state.pop();
+        }
+      }
+
+      if (char === `"`) {
+        if (state[state.length - 1] === 'expression' || state[state.length - 1] === 'single') {
+          state.push('double');
+        } else {
+          state.pop();
+        }
+      }
+
+      if (char === '{' && i >= 1 && state[state.length - 1] !== 'single' && state[state.length - 1] !== 'double') {
+        if (lineContent.charAt(i - 1) === '@') {
+          state.push('expression');
+        }
+      }
+
+      if (char === '}' && state[state.length - 1] === 'expression') {
+        state.pop();
+      }
+      i = i + 1;
+    }
+    finalState = state[state.length - 1];
+    return { matched: flag, state: finalState };
+  }
+
+  protected completion(params: TextDocumentPositionParams): Thenable<CompletionList | null> {
+    const document = this.documents.get(params.textDocument.uri);
+    if (!document) {
+      return Promise.resolve(null);
+    }
+    const text = this.getLGDocumentContent(document);
+    let templates: LGTemplate[] = [];
+    try {
+      const lgResources = getLGResources(text);
+      templates = lgResources.Templates;
+      // eslint-disable-next-line no-empty
+    } catch (error) {}
+    const completionList: CompletionItem[] = [];
+    templates.forEach(template => {
+      const item = {
+        label: template.Name,
+        kind: CompletionItemKind.Reference,
+        insertText:
+          template.Parameters.length > 0 ? template.Name + '(' + template.Parameters.join(', ') + ')' : template.Name,
+        documentation: template.Body,
+      };
+      completionList.push(item);
+    });
+
+    buildInfunctionsMap.forEach((value, key) => {
+      const item = {
+        label: key,
+        kind: CompletionItemKind.Function,
+        insertText: key + '(' + this.removeParamFormat(value.Params.toString()) + ')',
+        documentation: value.Introduction,
+      };
+      completionList.push(item);
+    });
+
+    const matchResult = this.matchedStates(params);
+    // TODO: more precise match
+    if (
+      matchResult &&
+      matchResult.matched &&
+      matchResult.state &&
+      allowedCompletionStates.includes(matchResult.state.toLowerCase())
+    ) {
+      return Promise.resolve({ isIncomplete: true, items: completionList });
+    } else {
+      return Promise.resolve(null);
+    }
+  }
+
+  protected validate(document: TextDocument): void {
+    this.cleanPendingValidation(document);
+    setTimeout(() => {
```

Comment:
A zero-delay timeout seems like a hack.

Reply:
here doValidate is designed be able to cancel, fixed